share/wake/lib/core/list.wake

```diff
@@ -499,51 +729,69 @@ export def tab (f: Integer => a): Integer => List a =
             loop (f (n-1), a) (n-1)
     loop Nil
 
-# seq: create a List of specified length with value ascending from 0.
+# Create a List of specified length with value ascending from 0.
 # For x <  y: at x (seq y) = Some x
 # For x >= y: at x (seq y) = None
 #
+# Example
+#   ```
 #   seq 10 = 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, Nil
 #   seq (-1) = Nil
+#   seq 0 = Nil
+#   ```
 export def seq: Integer => List Integer =
     tab (_)
 
-# zip: take two Lists and turn them into a List of Pairs
+# Take two Lists and turn them into a List of Pairs
 # The shortest length of the two input lists sets the output length.
 #
+# Example:
+#   ```
 #   zip (4, 7, Nil) ("a", "b", Nil) = (Pair 4 "a", Pair 7 "b", Nil)
+#   zip (4, 7, Nil) ("a", "b", "c", Nil) = (Pair 4 "a", Pair 7 "b", Nil)
+#   zip Nil x = Nil
+#   zip x Nil = Nil
+#   ```
 export def zip (a: List a) (b: List b): List (Pair a b) = match a b
     Nil _   = Nil
     _   Nil = Nil
     (lh,lt) (rh,rt) = Pair lh rh, zip lt rt
 
-# unzip: turn a List of Pairs into a Pair of Lists
+# Turn a List of Pairs into a Pair of Lists
 #
+# Guarantees:
+#  - Both output lists will be the same length
+#  - if `(len x) == (len y)` then `unzip (zip x y) = Pair x y`
+#  - `(\(Pair x y) zip x y) (unzip l) = l`
+#
+# Example:
+#   ```
 #   unzip (Pair 4 "a", Pair 7 "b", Nil) = Pair (4, 7, Nil) ("a", "b", Nil)
+#   unzip Nil = Pair Nil Nil
+#   ```
```

Comment:
In retrospect, I wonder if anyone actually uses `zip`/`unzip`. I added them simply because `list.wake` was modeled on the scala `Seq` type. Giving `Pair` special treatment seems rather odd, IMO.

Reply:
I'm down to delete unzip honestly, let me know what you think (@ag-eitilt as well). I've seldom if ever had a reason to use it. When it comes up you can use a pair of maps for it instead I think.

Regardless lets maybe remove unzip in a different PR

Reply:
Definitely a separate PR.  I agree that `zip`/`unzip` as part of the "List API" doesn't make as much sense (the main ways I use them in Haskell involve either `uncurry` or infinite lists, neither of which we have here), but it might fit a bit better as part of the "Pair API" -- conceptualized not as a special-casing lists which happen to contain pairs, but as an operation on pairs gathered in some collection.  Still probably not something we need to keep.

Reply:
I use zip a fair but but not unzip. Also zipWith is way cooler than zip anyway and we seem to be lacking it. zip is just zipWith Pair then.